src/fparser/two/Fortran2003.py

```diff
@@ -7203,102 +7208,417 @@ def match(string):
-                    Format_Item(repmap(line[i+1:].lstrip()))
-        for p in '/:':
-            if p in line:
-                l, r = line.split(p, 1)
-                return Format_Item(repmap(l.rstrip())), \
-                    Format_Item(p+repmap(r.lstrip()))
+        '''Implements the matching for the C1002 Format Item constraint. The
+        constraints specify certain combinations of format items that
+        do not need a comma to separate them. Rather than sorting this
+        out when parsing the list, it was decided to treat these as
+        individual and match them in this class. As a result the
+        generated class hierarchy is a little more complicated.
+
+        :param str string: The string to check for conformance with a \
+                           C1002 format item constraint.
+        :return: `None` if there is no match, otherwise a tuple of \
+        size 2 containing a mixture of Control_Edit_Descriptor and \
+        Format_Item classes depending on what has been matched.
+
+        :rtype: `NoneType` or ( \
+        :py:class:`fparser.two.Control_Edit_Desc`, \
+        :py:class:`fparser.two.Format_Item` ) or \
+        (:py:class:`fparser.two.Format_Item`, \
+        :py:class:`fparser.two.Control_Edit_Desc`) or \
+        (:py:class:`fparser.two.Format_Item`, \
+        :py:class:`fparser.two.Format_Item`)
+
+        '''
+        if not string:
+            return None
+        strip_string = string.strip()
+        if len(strip_string) <= 1:
+            return None
+        if strip_string[0] in ':/':
+            return Control_Edit_Desc(strip_string[0]), \
+                Format_Item(strip_string[1:].lstrip())
+        if strip_string[-1] in ':/':
+            return Format_Item(strip_string[:-1].rstrip()), \
+                Control_Edit_Desc(strip_string[-1])
+        line, repmap = string_replace_map(strip_string)
+        index = 0
+        while index < len(line) and line[index].isdigit():
+            index += 1
+        if index:
+            result = line[index].upper()
+            if result == '/':
+                return Control_Edit_Desc(repmap(line[:index+1])), \
+                    Format_Item(repmap(line[index+1:].lstrip()))
+            elif result == 'P':
+                # Rule C1002 only allows a comma to be ommited between
+                # a P edit descriptor and a following F, E, EN, ES, D,
+                # or G edit descriptor with an optional repeat
+                # specifier. In fparser2 this translates to a
+                # Format_Item instance containing a Data_Edit_Desc, or
+                # Data_Edit_Desc_C1002 instance as its second item
+                # with the data edit descriptor instance's first item
+                # specifying the type of edit descriptor.
+                lhs = Control_Edit_Desc(repmap(line[:index+1]))
+                rhs = Format_Item(repmap(line[index+1:].lstrip()))
+                if not isinstance(rhs, Format_Item):
```

Comment:
Format_Item is not really returning anything, rather it should contain (within its items method) an instance of Data_Edit_Descriptor within its items. I'm not sure about what other valid content it could have but I'm not sure that is important. Have I missed something?